src/databricks/labs/ucx/azure/azure_credentials.py

```diff
@@ -0,0 +1,283 @@
+                f"Will not reuse this client_secret"
+            )
+            return None
+
+    def _fetch_client_secret(self, sp_list: list[StoragePermissionMapping]) -> list[ServicePrincipalMigrationInfo]:
+        # check AzureServicePrincipalInfo from AzureServicePrincipalCrawler, if AzureServicePrincipalInfo
+        # has secret_scope and secret_key not empty, fetch the client_secret and put it to the client_secret field
+        #
+        # The input StoragePermissionMapping may have managed identity mixed in, we will ignore them for now, as
+        # they won't have any client_secret, we will process managed identity in the future.
+
+        # fetch client_secrets of crawled service principal, if any
+        azure_sp_info_with_client_secret: dict[str, str] = {}
+        azure_sp_infos = self._azure_sp_crawler.snapshot()
+
+        for azure_sp_info in azure_sp_infos:
+            if not azure_sp_info.secret_scope:
+                continue
+            if not azure_sp_info.secret_key:
+                continue
+            secret_value = self._read_databricks_secret(
+                azure_sp_info.secret_scope, azure_sp_info.secret_key, azure_sp_info.application_id
+            )
+            if secret_value:
+                azure_sp_info_with_client_secret[azure_sp_info.application_id] = secret_value
+
+        # update the list of ServicePrincipalMigrationInfo if client_secret is found
+        sp_list_with_secret = []
+        for sp in sp_list:
+            if sp.client_id in azure_sp_info_with_client_secret:
+                sp_list_with_secret.append(
+                    ServicePrincipalMigrationInfo(sp, azure_sp_info_with_client_secret[sp.client_id])
+                )
+        return sp_list_with_secret
+
+    def _print_action_plan(self, sp_list_with_secret: list[ServicePrincipalMigrationInfo]):
+        # print action plan to console for customer to review.
+        for sp in sp_list_with_secret:
+            print(
+                f"Service Principal name: {sp.service_principal.principal}, "
+                f"application_id: {sp.service_principal.client_id}, "
+                f"privilege {sp.service_principal.privilege} "
+                f"on location {sp.service_principal.prefix}"
+            )
+
+    def _generate_migration_list(self):
+        """
+        Create the list of SP that need to be migrated, output an action plan as a csv file for users to confirm
+        """
+        # load sp list from azure_storage_account_info.csv
+        sp_list = self._azure_resource_permissions.load()
+        # list existed storage credentials
+        sc_set = self._list_storage_credentials()
+        # check if the sp is already used in UC storage credential
+        filtered_sp_list = [sp for sp in sp_list if sp.client_id not in sc_set]
+        # fetch sp client_secret if any
+        sp_list_with_secret = self._fetch_client_secret(filtered_sp_list)
+        self._final_sp_list = sp_list_with_secret
```

Comment:
Can you rather return this as a result instead of modifying a shared state?
```suggestion
        return sp_list_with_secret
```

Reply:
fixed